Fixed #15665 -- Fixed inlines on a ModelAdmin with an inline to a child related model.

olivierdalang opened the pull request:
https://github.com/django/django/pull/6818 was closed due to inactivity. Here it is again, with Tim's comment included.

timgraham commented on tests/admin_inlines/models.py:
```diff
@@ -174,6 +179,7 @@ class FootNote(models.Model):
     chapter = models.ForeignKey(Chapter, models.PROTECT)
     note = models.CharField(max_length=40)
 
+
```
Please revert these unrelated whitespace changes.

timgraham commented on tests/admin_inlines/models.py:
```diff
@@ -252,9 +258,8 @@ class SomeChildModel(models.Model):
     position = models.PositiveIntegerField()
     parent = models.ForeignKey(SomeParentModel, models.CASCADE)
 
-# Other models
```
Please revert these unrelated whitespace changes.

olivierdalang replied:
I reverted those, but now the flake8 tests don't pass.

timgraham replied:
Looking at the diff, there are still unrelated whitespace changes.

timgraham commented on django/contrib/admin/helpers.py:
```diff
@@ -348,8 +348,10 @@ def __iter__(self):
     def needs_explicit_pk_field(self):
-        # Auto fields are editable (oddly), so need to check for auto or non-editable pk
-        if self.form._meta.model._meta.has_auto_field or not self.form._meta.model._meta.pk.editable:
+        # If the primary key is not editable, the form must provide a pk field explicitely.
```
Wrap comments at 79 characters.

timgraham commented on django/contrib/admin/helpers.py:
```diff
@@ -348,8 +348,10 @@ def __iter__(self):
-        # Auto fields are editable (oddly), so need to check for auto or non-editable pk
-        if self.form._meta.model._meta.has_auto_field or not self.form._meta.model._meta.pk.editable:
+        # If the primary key is not editable, the form must provide a pk field explicitely.
+        # Oddly, AutoField and One-to-one (for child models) are considered as editable, so we check for these two case
```
chop "we". Explain "Oddly".

timgraham commented on tests/admin_inlines/models.py:
```diff
@@ -63,6 +63,11 @@ def save(self, *args, **kwargs):
         super(NonAutoPKBook, self).save(*args, **kwargs)
 
 
+# Models for #15665
```
no need for the comment

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -906,3 +906,42 @@ def test_collapsed_inlines(self):
+
+    def test_non_autofield_primary_key(self):
+        """
+        Test for #15665 where inline wouldn't work if using non-AutoField ids and inherinting the model
```
State the expected behavior rather than "Test ...." (all tests test things).
No need for a selenium test as no JavaScript is involved. See 3aad955ea8db1592fad0012155eaa25b72e50dc5 for similar changes.

timgraham commented on tests/admin_inlines/admin.py:
```diff
@@ -28,6 +28,11 @@ class NonAutoPKBookStackedInline(admin.StackedInline):
 
+class NonAutoPKBookSubclassStackedInline(admin.StackedInline):
+    model = NonAutoPKBookSubclass
+    classes = ('collapse',)
```
Unneeded, I think.

olivierdalang commented:
Ok I made the changes you asked for. Let me know if there's anything else.

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -361,6 +361,24 @@ def test_inline_nonauto_noneditable_pk(self):
+        # with a non AutoField PK (as described in issue #15665).
+        """
+
+        author = Author.objects.create(name="Foo")
```
prefer single quotes unless the string contains a double quote

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -361,6 +361,24 @@ def test_inline_nonauto_noneditable_pk(self):
+        Ensure that the inline form work when the inner model inherits a model
+        # with a non AutoField PK (as described in issue #15665).
+        """
+
```
No blank lines needed (try to follow the style of nearby tests)

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -361,6 +361,24 @@ def test_inline_nonauto_noneditable_pk(self):
 
+    def test_non_autofield_primary_key(self):
+        """
+        Ensure that the inline form work when the inner model inherits a model
```
Try to state the expected behavior rather than "Ensure that" or "Test that" since all tests do that. Ticket references are only needed for obscure issues that can't be explained in the docstring. I'm not sure that's the case here.

timgraham commented on tests/admin_inlines/tests.py:
```diff
@@ -361,6 +361,24 @@ def test_inline_nonauto_noneditable_pk(self):
+        """
+        author = Author.objects.create(name='Foo')
+        book1 = NonAutoPKBookSubclass.objects.create(title='Bar', author=author)
+
```
please chop the rest of the blank lines

olivierdalang commented:
Removed the blank lines and squashed.
Looking forward to pop the champagne for my first contribution merged into Django :) Thanks for the patience and the detailed review.

timgraham commented:
It seems this doesn't entirely resolve the ticket. The point about that it "leads to wrong behavior, when modifying the value of the concrete base class's primary key (the name field in the example above) in the inline admin, resulting in a new entry." still looks relevant. Did you consider that case?

olivierdalang commented:
@timgraham No indeed I did not consider this case, but only the simpler case where the PK is not editable.

Using an user-editable property as primary key in a model with inheritance sounds like a bad idea anyway.

Do you think it is acceptable to merge as a partial fix ?

timgraham commented:
This seems to introduce a sort of data corruption for the case I mentioned with extra objects being created (whereas right now things simply crash to prevent any edits), so I don't think it's acceptable if that understanding is correct.

timgraham commented:
Closing due to inactivity. Please send a new PR if you return to this.
